components/datetime/src/format/datetime.rs

```diff
@@ -247,289 +223,217 @@ where
     DS: DateSymbols<'data>,
     TS: TimeSymbols,
 {
-    let mut r = Ok(());
-
-    match field.symbol {
-        FieldSymbol::Era => {
-            let era = datetime
-                .datetime()
-                .year()
-                .ok_or(Error::MissingInputField(Some("year")))
-                .map(|x| x.era);
-
-            let symbol = era.as_ref().map_err(|&e| e).and_then(|era| {
-                Ok(date_symbols
-                    .ok_or(Error::MissingDateSymbols)?
-                    .get_symbol_for_era(field.length, era)?
-                    .unwrap_or(&era.0))
-            });
+    // Writes an error string for the current symbol
+    let write_value_missing = |w: &mut W| {
+        w.with_part(Part::ERROR, |w| {
+            "{".write_to(w)?;
+            char::from(field.symbol).write_to(w)?;
+            "}".write_to(w)
+        })
+    };
 
-            let () = match symbol {
-                Ok(s) => w.write_str(s)?,
-                Err(e) => {
-                    r = r.and(Err(e));
-                    w.with_part(Part::ERROR, |w| "<ICU4X ERROR>".write_to(w))?
-                }
-            };
-        }
-        FieldSymbol::Year(year) => match year {
-            Year::Calendar => {
-                let () = match datetime
-                    .datetime()
-                    .year()
-                    .ok_or(Error::MissingInputField(Some("year")))
-                    .map(|x| x.number)
-                    .map(FixedDecimal::from)
-                {
-                    Ok(x) => r = r.and(try_write_number(w, fixed_decimal_format, x, field.length)?),
+    Ok(match (field.symbol, field.length) {
+        (FieldSymbol::Era, l) => {
+            if let Some(year) = datetime.datetime().year() {
+                match date_symbols
+                    .ok_or(Error::MissingDateSymbols)
+                    .and_then(|ds| {
+                        ds.get_symbol_for_era(l, &year.era)
+                            .map(|e| e.unwrap_or(&year.era.0))
+                    }) {
```

Comment:
Nit: I think you don't need the extra `map` here. It should use the `Err` path in the match

```suggestion
                        ds.get_symbol_for_era(l, &year.era)
                    }) {
```

Reply:
See the test I had to change. It was a mechanism to write out the era name fallback. TryWriteable is a better solution. Hmm, I think this trait is internal, so we could probably change it to return `Result<&str>`.

Reply:
Will address this in a follow-up